File: day2/day2.py
def calculate(input_array: list, aim_enabled: bool):

    depth = 0
    horizontal = 0
    aim = 0

    for dir, amt in input_array:
        if dir == 'forward':
            horizontal += amt
            if aim_enabled:
                depth += aim * amt
        elif dir == 'up':
            if aim_enabled:
                aim -= amt
            else:
                depth -= amt
        elif dir == 'down':
            if aim_enabled:
                aim += amt
            else:
                depth += amt

    return horizontal * depth

with open('input.txt', 'r') as file:
    input_data = [(x, int(y)) for x, y in [x.strip().split(' ') for x in file.readlines()]]


print(calculate(input_data, False))
print(calculate(input_data, True))

File: AoC21/day2/day2.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <regex>
#include <bits/stdc++.h>

using namespace std;

struct Instruction {
    string direction;
    int amount;
};

vector<string> split(string str, string splitter) {
    vector<string> result;

    int start = 0;
    int end = str.find(splitter);

    while (end != - 1) {
        result.push_back(str.substr(start, end - start));
        start = end + splitter.size();
        end = str.find(splitter, start);
    }

    result.push_back(str.substr(start, end - start));

    return result;
}

vector<Instruction> readInput() {
    string line;
    int val;
    string direction;
    vector<string> values;
    string regex_str = " ";
    vector<Instruction> inputData;

    ifstream inputFile("input.txt");

    if (inputFile.is_open()) {
        while (getline(inputFile, line)) {
            values = split(line, regex_str);

            direction = values[0];
            val = std::stoi(values[1]);

            Instruction i{.direction = direction, .amount = val};

            inputData.push_back(i);
        }
        inputFile.close();
    }

    return inputData;
}

void part1(vector<Instruction> instructions) {
    int horizontal = 0;
    int depth = 0;
    int result;
    string dir;

    for (auto instr : instructions) {

        dir = instr.direction;

        if (dir == "forward") {
            horizontal += instr.amount;
        } else if (dir == "up") {
            depth -= instr.amount;
        } else if (dir == "down") {
            depth += instr.amount;
        }
    }

    result = horizontal * depth;

    cout << result << endl;
}

int main() {
    vector<Instruction> inputs;

    inputs = readInput();

    part1(inputs);

    return 0;
}
